### PR TITLE
Pynq-fud integration

### LINKED ISSUE #1114: Integrate PYNQ into fud stage
As mentioned here:

> I think there's one thing that would be good for quality of life: making this all work in fud. I think so far you've been manually invoking your PYNQ host program to run a given `.xclbin`; is that right? It shouldn't be _too_ much effort to integrate this back into fud, just by making fud execute your PYNQ driver script as a subprocess. This would make it super easy to rerun stuff when doing item 3 above. As a bonus, this would address #1037 by automatically setting up the environment variables for that host-code execution.

_Originally posted by @sampsyo in https://github.com/cucapra/calyx/discussions/1022#discussioncomment-3175448_

Hopefully, this will also address some combination of:

- [ ] #1037 
- [ ] #872 
- [ ] #856

Although maybe not completely

### PR DESCRIPTION
**Pull request by nathanielnrn**
This PR can close #1114.
A new pynq-fud-script can take in arbitrary data, run a kernel on it, and output the relevant memories. Integrates into `fud e foo.xclbin --from xclbin --to fpga ...` commands.

This replaces the old OpenCL way of running a kernel that fud used to use.

Unfortunately this does __not__ address #1037 #872 #856.

After this, I plan on working on #1104, unless @rachitnigam or @sampsyo thinks I should try to first quickly close the unaddressed issues mentioned.

### CODE REVIEW

File: fud/fud/stages/xilinx/execution.py [hunk 1]
```diff
@@ -33,26 +35,15 @@ def _define_steps(self, input, builder, config):
...
-        
-        @builder.step()
-        def import_libs():
-            """Import optional libraries"""
-            try:
-                import pyopencl as cl  # type: ignore
-
-                self.cl = cl
-            except ImportError:
-                raise errors.RemoteLibsNotInstalled
```
Comment (nathanielnrn) [inline fud/fud/stages/xilinx/execution.py:45]: Wasn't sure about this, should the pynq import from `pyproject.toml` go here instead?

File: fud/fud/stages/xilinx/execution.py [hunk 2]
```diff
@@ -71,65 +62,39 @@ def run(xclbin: SourceType.Path) -> SourceType.String:
...
-                    f.writelines([
-                        "open_vcd\n",
-                        "log_vcd *\n",
-                    ])
+                    f.writelines(
+                        [
+                            "open_vcd\n",
+                            "log_vcd *\n",
+                        ]
+                    )
```
Comment (rachitnigam) [inline fud/fud/stages/xilinx/execution.py:81]: Move formatting commits into a different PR

File: fud/pyproject.toml [hunk 3]
```diff
@@ -19,7 +19,8 @@ requires = [
   "packaging",
   "numpy",
   "simplejson",
-  "networkx>=2.6"
+  "networkx>=2.6",
+  "pynq"
```
Comment (rachitnigam) [inline fud/pyproject.toml:23]: Move this into the `extra.fpga` requirements so that it isn't installed by default

File: fud/fud/stages/xilinx/execution.py (hunk 1 above)
Comment (rachitnigam) [reply fud/fud/stages/xilinx/execution.py:45]: The reason this is inside a `try`-`catch` inside a function is because the `fpga` dependencies are not installed by default. The program only throws an error if you try to run the `xilinx` stage without installing the `fpga` dependencies in the `flit` configuration.

I would recommend moving the
```
from fud.stages.xilinx import fud_pynq_script
```
and using a `self.pynq_script` field to access it instead of importing it on the top which will cause errors for default `fud` installations

File: fud/pyproject.toml (hunk 3 above)
Comment (rachitnigam) [reply fud/pyproject.toml:23]: Also, if `pyopencl` is no longer used, remove it from the dependencies below
